### PR TITLE
Add synchronization priorities on each Bluetooth Service

### PR DESCRIPTION
**Pull request by xavier-fernandez**
- This commit adds a way of synchronizing bluetooth services in an
  specific order.

### CODE REVIEW

File: lib/src/androidTest/java/com/sensirion/libble/devices/BleDeviceTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import android.test.suitebuilder.annotation.SmallTest;
+
+import com.sensirion.libble.services.BleService;
+import com.sensirion.libble.services.MockBleService;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static com.sensirion.libble.services.BleServiceSynchronizationPriority.HIGH_PRIORITY;
```
Comment (abrauchli) [inline lib/src/androidTest/java/com/sensirion/libble/devices/BleDeviceTest.java]: (applies to every static import of this commit) import the enum, not each enum value individually and use it with the enum's name to give context: BleServiceSynchronizationPriority.HIGH_PRIORITY

In same places you're already doing it like this. Let's be consistent

File: lib/src/main/java/com/sensirion/libble/devices/BleDevice.java [hunk 2]
```diff
@@ -164,4 +177,25 @@
...
+    /**
+     * Comparator for comparing the different {@link BleService} using their
+     * {@link BleServiceSynchronizationPriority} for comparing them.
+     */
+    @NonNull
+    Comparator<BleService> SERVICE_PRIORITY_COMPARATOR = new Comparator<BleService>() {
+        @Override
+        public int compare(@NonNull final BleService serviceA,
+                           @NonNull final BleService serviceB) {
+            switch (serviceA.getServiceSynchronizationPriority()) {
```
Comment (abrauchli) [inline lib/src/main/java/com/sensirion/libble/devices/BleDevice.java]: that whole method looks a lot more complicated than it should be:
return serviceB.getServiceSynchronizationPriority().ordinal() - serviceA.getServiceSynchronizationPriority().ordinal()

File: lib/src/main/java/com/sensirion/libble/devices/Peripheral.java [hunk 3]
```diff
@@ -741,6 +741,27 @@ public boolean areAllServicesReady() {
      * {@inheritDoc}
      */
     @Override
+    public boolean synchronizeAllDeviceServices() {
+        final List<BleService> servicesToSynchronize = new LinkedList<>(mServices);
+        // Synchronize all the services considering their priorities
+        Collections.sort(servicesToSynchronize, SERVICE_PRIORITY_COMPARATOR);
+        for (final BleService service : servicesToSynchronize) {
+            if (service.isServiceReady()) {
```
Comment (abrauchli) [inline lib/src/main/java/com/sensirion/libble/devices/Peripheral.java]: no need for continue:

if (!service.isServiceReady()) {
  service.synchronizeService();
  return false;
}

File: lib/src/main/java/com/sensirion/libble/services/AbstractBleService.java [hunk 4]
```diff
@@ -326,4 +328,20 @@ public boolean equals(@Nullable final Object otherService) {
     public String toString() {
         return String.format("%s of the device: %s", getClass().getSimpleName(), getDeviceAddress());
     }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int compareTo(@NonNull final BleService otherService) {
+        switch (this.getServiceSynchronizationPriority()) {
```
Comment (abrauchli) [inline lib/src/main/java/com/sensirion/libble/services/AbstractBleService.java]: see comment above

**Comment by abrauchli** [comment]
See comments and questions. Also the commit message is missing a description to how priorities are assigned (and why this assignment was chosen), since that's also part of the commit.

File: lib/src/androidTest/java/com/sensirion/libble/devices/BleDeviceTest.java (hunk 1 above)
Comment (xavier-fernandez) [inline lib/src/androidTest/java/com/sensirion/libble/devices/BleDeviceTest.java]: done

File: lib/src/main/java/com/sensirion/libble/devices/BleDevice.java (hunk 2 above)
Comment (xavier-fernandez) [inline lib/src/main/java/com/sensirion/libble/devices/BleDevice.java]: Wow, did not know about this method. 👍 

Thank you

Comment (xavier-fernandez) [inline lib/src/main/java/com/sensirion/libble/devices/BleDevice.java]: Done

File: lib/src/main/java/com/sensirion/libble/devices/Peripheral.java (hunk 3 above)
Comment (xavier-fernandez) [inline lib/src/main/java/com/sensirion/libble/devices/Peripheral.java]: I did this for following the Sensirion convention: `Avoid using negative if conditions`

I can use negations on this type of cases on the future?

Comment (xavier-fernandez) [inline lib/src/main/java/com/sensirion/libble/devices/Peripheral.java]: done

File: lib/src/main/java/com/sensirion/libble/services/AbstractBleService.java (hunk 4 above)
Comment (xavier-fernandez) [inline lib/src/main/java/com/sensirion/libble/services/AbstractBleService.java]: done

**Comment by xavier-fernandez** [comment]
@abrauchli Did an amend fixing your code remarks

**Comment by xavier-fernandez** [comment]
Did an amend that fixes the remarks, adds a missing test `BleServiceTest.compareTo()` and updates the commit message. 
 cce6958...d581310

**Comment by abrauchli** [comment]
this seems to be part of PR #13 - let's close this then

**Pull request closed without merging**